packages/storage/src/postgres.rs

```diff
@@ -678,6 +685,134 @@ impl Postgres {
         Ok(())
     }
 
+    pub(crate) async fn _update_costs(
+        &self,
+        cost_per_tx: Vec<([u8; 32], u128, u64)>,
+    ) -> Result<()> {
+        let mut tx = self.connection_pool.begin().await?;
+
+        // A map to accumulate cost and size contributions per bundle
+        let mut bundle_updates: HashMap<i32, BundleCostUpdate> = HashMap::new();
+
+        for (hash, total_fee, da_block_height) in cost_per_tx {
+            let row = sqlx::query!(
+                r#"
+                SELECT
+                    f.bundle_id,
+                    SUM(f.total_bytes)::BIGINT AS total_bytes,
+                    SUM(f.unused_bytes)::BIGINT AS unused_bytes
+                FROM
+                    l1_blob_transaction t
+                    JOIN l1_transaction_fragments tf ON t.id = tf.transaction_id
+                    JOIN l1_fragments f ON tf.fragment_id = f.id
+                WHERE
+                    t.hash = $1
+                GROUP BY
+                    f.bundle_id
+                "#,
+                hash.as_slice()
+            )
+            .fetch_one(&mut *tx)
+            .await?;
+
+            let bundle_id = row.bundle_id;
+            let total_bytes: i64 = row.total_bytes.unwrap_or(0);
+            let unused_bytes: i64 = row.unused_bytes.unwrap_or(0);
+            let size_contribution = (total_bytes - unused_bytes) as u64;
+
+            // Accumulate contributions in the bundle_updates map
+            let entry = bundle_updates.entry(bundle_id).or_insert(BundleCostUpdate {
+                cost_contribution: 0,
+                size_contribution: 0,
+                latest_da_block_height: 0,
+            });
+
+            entry.cost_contribution += total_fee;
+            entry.size_contribution += size_contribution;
+            entry.latest_da_block_height = da_block_height; // Update with the latest da_block_height
+        }
+
+        // update the bundle_cost table for each affected bundle
+        for (bundle_id, update) in bundle_updates {
+            // check if any fragment in the bundle is not associated with a finalized transaction
+            let is_finalized = sqlx::query!(
+                r#"
+                SELECT COUNT(*) = 0 AS "is_finalized"
+                FROM l1_fragments f
+                LEFT JOIN l1_transaction_fragments tf ON f.id = tf.fragment_id
```

Comment:
This query will flag unsubmitted fragments as finalized:

1. have an unsubmitted fragment (no tx for it)
2. left join on transaction fragments, you get NULL for all the TF fields
3. left join on l1 blob transactions you get NULL for all the T fields
4. t.state (NULL) is distinct from Finalized.

Reply:
In that case COUNT(*)  = 0 evaluates to false